dist/sizzle.js

```diff
@@ -1628,13 +1633,237 @@ function setFilters() {}
+		"ATTR": 70,
+		"PSEUDO": 60
+	};
+
+	var POSITIONAL_PSEUDOS = [
+		"nth", "first", "last", "eq", "even", "odd", "lt", "gt", "not"
+	];
+
+	/** 
+	 * A function that defines the sort order.
+	 * Returns a value lesser than 0 if "left" is less than "right".
+	 */
+	var compareFunction = function(left, right) {
+		var leftValue = TOKEN_TYPES_VALUES[left.type];
+		var rightValue = TOKEN_TYPES_VALUES[right.type];
+		return leftValue - rightValue;
+	};
+
+	/**
+	 * Checks if the specified tokens group is sortable.
+	 * We do not re-sort tokens in case of any positional or child pseudos in the group
+	 */
+	var isSortable = function(tokens) {
+
+		var iTokens = tokens.length;
+		while (iTokens--) {
+			var token = tokens[iTokens];
+			if (token.type === "PSEUDO" && POSITIONAL_PSEUDOS.indexOf(token.matches[0]) !== -1) {
+				return false;
+			}
+			if (token.type === "CHILD") {
+				return false;
+			}
+		}
+		return true;
+	};
+
+	/**
+	 * Sorts the tokens in order to mitigate the issues caused by the left-to-right matching.
+	 * The idea is change the tokens order so that Sizzle was matching fast selectors first (id, class),
+	 * and slow selectors after that (and here I mean our slow custom pseudo classes).
+	 * 
+	 * @param {Array} tokens An array of tokens to sort
+	 * @returns {Array} A new re-sorted array
+	 */
+	var sortTokens = function(tokens) {
+
+		if (!tokens || tokens.length === 1) {
+			return tokens;
+		}
+
+		var sortedTokens = [];
+		var groups = splitCompoundSelector(tokens);
+		for (var i = 0; i < groups.length; i++) {
+			var group = groups[i];
+			if (group instanceof Array) {
+				if (isSortable(group)) {
+					group.sort(compareFunction);
+				}
+				sortedTokens = sortedTokens.concat(group);
+			} else {
+				sortedTokens.push(group);
+			}
+		}
+
+		return sortedTokens;
+	};
+
+	/**
+	 * Sorts every tokens array inside of the specified "groups" array.
+	 * See "sortTokens" methods for more information on how tokens are sorted.
+	 * 
+	 * @param {Array} groups An array of tokens arrays.
+	 * @returns {Array} A new array that consists of the same tokens arrays after sorting
+	 */
+	var sort = function(groups) {
+		var sortedGroups = [];
+		var len = groups.length;
+		var i = 0;
+		for (; i < len; i++) {
+			sortedGroups.push(sortTokens(groups[i]));
+		}
+		return sortedGroups;
+	};
+
+	// Expose
+	return {
+		sort: sort
+	};
+})();
+
+/**
+ * [AdGuard Patch]:
+ * Removes trailing spaces from the tokens list
+ * 
+ * @param {*} tokens An array of Sizzle tokens to post-process
+ */
+function removeTrailingSpaces(tokens) {
+	var iTokens = tokens.length;
+	while (iTokens--) {
+		var token = tokens[iTokens];
+		if (token.type === " ") {
+			tokens.length = iTokens;
+		} else {
+			break;
+		}
+	}
+}
+
+/**
+ * [AdGuard Patch]:
+ * 
+ * This method processes parsed token groups, divides them into a number of selectors
+ * and makes sure that each selector's tokens are cached properly in Sizzle.
+ * 
+ * @typedef {Object} SelectorData
```

Comment:
This can be made more precise, otherwise what's the point of having a `typedef`?

Reply:
What do you mean? This is a precise type definition: two `@property` that follow `@typedef` describe this type.

Reply:
I meant something like this: 
```js
/** @typedef {{ groups: Array<Array<?>>, selectorText: string }} SelectorData */
```

Reply:
http://usejsdoc.org/tags-typedef.html

Reply:
Are you saying that the above is an invalid jsdoc?

Reply:
Both are valid, mine is more precise.

Reply:
I don't think so. Please stay in point, what are you trying to say with a link to usejsdoc.org?

Reply:
https://uploads.adguard.com/up04_o8u8r_sizzle-extcsssizzle.js_at_init__AdguardTeamsizzle-extcss.png

Reply:
ok, I was too accustomed to Closure-compiler style JSDoc :) It is meaningless in closure compiler and honestly I prefer it: https://groups.google.com/forum/#!topic/closure-compiler-discuss/vZNIh9vkJPU

Reply:
I prefer describing properties as well and the closure-compiler's style does not let me do it.

Now we're both happy:
https://github.com/AdguardTeam/sizzle-extcss/pull/1/commits/57d2695f972d98377bf3667005490340c2fe7fc4